Set-DomainUser.ps1

```diff
@@ -971,375 +1274,234 @@ if (!($NETBIOS)) {
+        Write-LogMessage -Type Verbose -MSG "PSMConnect user credentials validated"
     }
     else {
-        Write-Error "PSMConnect user validation failed. Please validate PSMConnect user name and password or remove -TestPsmConnectCredentials to skip this test"
+        Write-LogMessage -Type Error -MSG "PSMConnect user validation failed. Please validate PSMConnect user name and password or remove -TestPsmConnectCredentials to skip this test"
         exit 1
     }
 }
-    
+
 if ($TestPsmAdminCredentials) {
     if (ValidateCredentials -domain $domain -Credential $psmAdminCredentials) {
-        Write-Host "PSMAdminConnect user credentials validated"
+        Write-LogMessage -Type Verbose -MSG "PSMAdminConnect user credentials validated"
     }
     else {
-        Write-Error "PSMAdminConnect user validation failed. Please validate PSMConnect user name and password or remove -TestPsmAdminCredentials to skip this test."
+        Write-LogMessage -Type Error -MSG "PSMAdminConnect user validation failed. Please validate PSMConnect user name and password or remove -TestPsmAdminCredentials to skip this test."
         exit 1
     }
 }
-    
+
 # Reverse logic on script invocation setting because double negatives suck
 $DoHardening = !$DoNotHarden
 $DoConfigureAppLocker = !$DoNotConfigureAppLocker
-    
+
 Write-Host "Logging in to CyberArk"
-$pvwaToken = New-ConnectionToRestAPI -pvwaAddress $pvwaAddress -tinaCreds $tinaCreds
-if (Test-PvwaToken -Token $pvwaToken -pvwaAddress $pvwaAddress) {
-    Write-Host "Successfully logged in"
+$pvwaTokenResponse = New-ConnectionToRestAPI -pvwaAddress $pvwaAddress -tinaCreds $tinaCreds
+if ($pvwaTokenResponse.ErrorCode -ne "Success") {
+    # ErrorCode will always be "Success" if Invoke-RestMethod got a 200 response from server.
+    # If it's anything else, it will have been caught by New-ConnectionToRestAPI error handler and an error response generated.
+    # The error message shown could be from a JSON response, e.g. wrong password, or a connection error.
+    Write-LogMessage -Type Error "Logon to PVWA failed. Result:"
+    Write-LogMessage -Type Error ("Error code: {0}" -f $pvwaTokenResponse.ErrorCode)
+    Write-LogMessage -Type Error ("Error message: {0}" -f $pvwaTokenResponse.ErrorMessage)
+    exit 1
+}
+if (!($pvwaTokenResponse.Response -match "[0-9a-zA-Z]{200,256}")) {
+    # If we get here, it means we got a 200 response from the server, but the data it returned was not a valid token.
+    # In this case, we display the response we got from the server to aid troubleshooting.
+    Write-LogMessage -Type Error "Response from server was not a valid token:"
+    Write-LogMessage -Type Error $pvwaTokenResponse.Response
+    exit 1
+}
+# If we get here, the token was retrieved successfully and looks valid. We'll still test it though.
+$PvwaTokenTestResponse = Test-PvwaToken -Token $pvwaTokenResponse.Response -pvwaAddress $pvwaAddress
+if ($PvwaTokenTestResponse.ErrorCode -eq "Success") {
```

Comment:
Should we use an actual code rather than a string?

Reply:
We can adjust if we feel the need. Doesn't really matter what we use as long as it's unambiguous and unique. It's never shown to the user, just used to check the result of the function.